### DISCUSSION TITLE
Issue with ESP8266 and DS18B20 sensor

### DISCUSSION THREAD

**Discussion by arthur-bastien** [micropython/micropython#9250] [Hardware & Peripherals] [answered]
Hello everyone,

First, I hope this is the good place to post my issue.
I'm a french guy with a small personnal project to create a homemade thermometer and as it seems to be a common project there is a lot of tutorials and forums and it'e easy to get informations.
I've found a lot of informations here so thank you everyone for your time and answers :) !

I started my project with a [french blog post](https://blog.betomorrow.com/iot-un-capteur-de-temp%C3%A9rature-connect%C3%A9-c7cf7073a3ba) wich describes how to set a [ESP8266 board](https://www.amazon.fr/gp/product/B01N9RXGHY/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1) with a [temperature sensor](https://www.amazon.fr/gp/product/B08G1C8MPQ/ref=ppx_yo_dt_b_asin_title_o06_s01?ie=UTF8&psc=1) and Arduino code. 
This worked very well, datas are sent to a external dashboard and I developed a small Android app to check values, values which seems accurates and corrects.

Lately [I found I could use MicroPython](https://randomnerdtutorials.com/micropython-ds18b20-esp32-esp8266/) to run my board and I was quite excited to try this as it appears to be easier to update code and run it, in particular with REPL prompt.
Firmware : esp8266-20220618-v1.19.1
I succeeded to code in MicroPython my program but there is an issue I could not solve although I read dozens of forums pages.

With the first DS18B20 sensor, I realised the values were weird : between 25°C and 33°C inside which are obviously not correct. 
Thanks to [cpetrich's work](https://github.com/cpetrich/counterfeit_DS18B20/), I founded my sensor was counterfeit and understand better the wrong values.
I ordered another [DS18B20 sensor from AZ-Delivery](https://www.amazon.fr/gp/product/B089QJ4TP5/ref=ppx_yo_dt_b_asin_title_o01_s00?ie=UTF8&psc=1) which arrived today, I hoped the values could be correct this time but this is not the case.
This sensor looks like an original one : lasered informations : 
```
DALLAS 
18B20
2034C4
+817AB
```

but discover_fake and classify_fake indicates this is counterfeit as you can see at the bottom.
Here is a sample of the temperature values every 5 seconds: 

```
esp8266 initialized
Found DS devices:  [bytearray(b'(\xa5\x16\x07\xd6\x01<\xde')]
Relevé de température :  25.0
Relevé de température :  27.0625
Relevé de température :  31.875
Relevé de température :  32.5625
Relevé de température :  32.75
Relevé de température :  33.0625
Relevé de température :  30.75
Relevé de température :  27.0625
```

I'm thinking about returning to Arduino code because it sounds like it doesn't care about counterfeit sensor or not but I would prefer finding a solution to my problem with MicroPython ;)

Could you please help me to find if this is only a hardware problem or if there is some improvements to do in my code ? 
Here is a public github repo with my code : https://github.com/arthur-bastien/public
Or if you have a link to a shop which sells original sensors with France shipping, feel free to share :) 

Thanks to everyone who read my post !
Have a good evening !

Under : discover_fake & classify_fake
```
discover_fake : 
19:24:09.617 -> 28-A5-16-07-D6-01-3C-DE: ROM does not follow expected pattern 28-xx-xx-xx-xx-00-00-crc. Error.
19:24:09.664 ->   Scratchpad Register: 90/01/55/05/7F/A5/81/66/4E
19:24:09.664 ->   Info only: Scratchpad bytes 2,3,4 (55/05/7F):  not Maxim default values 4B/46/7F.
19:24:09.664 ->   Scratchpad byte 5 (0xA5):  should have been 0xFF according to datasheet. Error.
19:24:09.664 ->   Scratchpad byte 6 (0x81):  unexpected value. Error.
19:24:09.664 ->   Scratchpad byte 7 (0x66):  should have been 0x10 according to datasheet. Error.
19:24:09.711 ->   0x4E modifies alarm registers:  ok.
19:24:09.711 ->   0x4E accepts 10 bit resolution:  ok.
19:24:09.711 ->   0x4E preserves reserved bytes:  ok.
19:24:09.711 ->   0x4E accepts 12 bit resolution:  ok.
19:24:09.711 ->   0x4E preserves reserved bytes:  ok.
19:24:09.711 ->   Checking byte 6 upon temperature change: not necessary. Skipped.
19:24:09.711 ->   --> Sensor appears to be counterfeit based on 4 deviations.

classify_fake :
19:21:21.592 -> classify_fake_DS18B20.ino version 2020/05/25
19:21:21.592 -> 
19:21:21.592 -> This sketch performs a minimal test to classify
19:21:21.592 ->   1-wire sensors attached to pin 2 according to
19:21:21.592 ->   https://github.com/cpetrich/counterfeit_DS18B20/
19:21:21.592 ->   assuming they belong to one of the DS18B20
19:21:21.639 ->   Families in circulation in 2019.
19:21:21.639 ->   The script evaluates the sensors based on their
19:21:21.639 ->   response to undocumented function codes.
19:21:21.639 ->   Proceed at your own risk!
19:21:21.639 -> 
19:21:21.639 ->   Make sure power pin is connected to either Vcc or GND.
19:21:21.639 ->   Hit enter to start analysis.
19:21:21.639 -> 
19:21:22.858 -> 28-A5-16-07-D6-01-3C-DE: Family D2 (Clone w/o parasitic power).
```

Attached : pictures from the sensor and the board.
![20220908_191445_resized](https://user-images.githubusercontent.com/100539710/189202679-0c1c5a22-14eb-47a4-ac55-6022f9737ee1.jpg)
![20220908_203407](https://user-images.githubusercontent.com/100539710/189202687-ccc365f6-f2fc-4481-a7d6-b8e38f98bbb6.jpg)

**Comment by davefes** [comment, marked as answer]
Try this between the convert and the read:

`        utime.sleep_ms(1000) #  min of 750ms for 1wire conversion`

**Reply by arthur-bastien** [reply]
Thanks for your quick answer :) 

I added your piece of code to my main.py ([see code](https://github.com/arthur-bastien/public/blob/main/main.py)).
The temperature seems more stable but is not good yet ; there are 8 extra degrees compared with a mercury thermometer.
```
esp8266 initialized
Found DS devices:  [bytearray(b'(\xa5\x16\x07\xd6\x01<\xde')]
Relevé de température :  32.625
Relevé de température :  33.5625
Relevé de température :  33.625
Relevé de température :  33.625
Relevé de température :  33.5625
Relevé de température :  33.5625
Relevé de température :  33.625
Relevé de température :  33.6875
Relevé de température :  33.75
```

**Reply by davefes** [reply]
With the Arduino code you did not see this variation?

+/- 0.1 C looks reasonable to me.

**Reply by davefes** [reply]
If you don't mind waiting 10 seconds take 10 readings and average them.  I take 100 samples for an ADC reading to filter out "noise", but that only takes 1 second.

**Reply by arthur-bastien** [reply]
I saw the variation with the Arduino code and yes, +/- 0.1 variation is reasonable but the temperature here is 24°C, not 33°C.
Even with 10 readings and average temperature is not really correct.

**Reply by davefes** [reply]
Ah, didn't notice that. Pull-up resistor value?  Maybe, calibrate in ice water.  Does the Arduino code read the correct temperature?

**Reply by arthur-bastien** [reply]
Resistor value is 4,7K Ω. I made some code modifications and put the board and sensor in the fridge with the mercury thermometer to check the values. 
Values are going down, slowly.

Arduino code read the correct temperature :
![image](https://user-images.githubusercontent.com/100539710/189232586-a51f3f37-89fc-47e5-9bf6-5d27872453fc.png)

I'll keep in touch tomorrow ;)

**Reply by robert-hh** [reply]
The test with my sensor and the esp8266 showed reasonable temperatures. I had earlier unexpected high temperatures after having touched the sensor, or when it was placed close to the warm object like the ESP8266 board or the vent of the computer. A false reading is not probable, due to the CRC check on the data.
We had other cases here that Arduino code behaved different that MicroPython ocde, just because the Arduino code does **not** start WiFi and thus the power consumption of the board is much lower, which causes a lower heat dissipation. Heat may also propagate through it's leads to the sensor.

**Reply by arthur-bastien** [reply]
I have experienced that touching the sensor raises the temperature that's why I just set 2 sensors on the board, [this one](https://www.amazon.fr/gp/product/B089QJ4TP5/ref=ppx_yo_dt_b_asin_title_o01_s00?ie=UTF8&psc=1) and the one with 1-meter wire. 
I also put the Arduino board's wire at the same place, both far from the laptop, and let them run for a while.

The Arduino board gives accurate datas (I think) with +/- 0.1°C variation every 2 seconds, which seems good to me.
The MicroPython board gives different values regarding the sensor and none are accurate. 

Updated [main.py](https://github.com/arthur-bastien/public/blob/main/main.py) to put thoses lines in the loop, is that correct ? 
```
ds_sensor.convert_temp()
utime.sleep_ms(1000)
temp=ds_sensor.read_temp(rom)
```

After a hour, the 2 sensors iwith a wire give the same value with 0.1°C of difference between each, that's totally acceptable for me.
I'll let them run a few days and update the discussion with results. 

@robert-hh : do you have a link about where you bought your sensor ? Maybe I should try with a genuine one :) 

Thanks to both of you for your help !

**Reply by robert-hh** [reply]
The script that I usually use for testing a port's implementation is:
```
import time
import machine
import onewire, ds18x20

dat = machine.Pin(16)

# create the onewire object
ds = ds18x20.DS18X20(onewire.OneWire(dat))

# scan for devices on the bus
roms = ds.scan()
print('found devices:', roms)

# loop 10 times and print all temperatures
for i in range(5):
    print('temperatures:', end=' ')
    ds.convert_temp()
    time.sleep_ms(750)
    for rom in roms:
        print(ds.read_temp(rom), end=' ')
    print()
```
As far as I recall, I purchased my sensors at a distributor called reichelt.de. There is a reichelt.com site as well. But the sensor looks pretty expensive there, like 3 times more than I paid for them a few years ago.

**Reply by robert-hh** [reply]
I just looked at mouser.com an digikey.com, and there the price is even higher, if available at all. Thus you can assume that devices sold by e.g. AZ-Delivery for less than 1 € are counterfeit.

**Reply by arthur-bastien** [reply]
Thanks a  lot for your research and your script, will try it.

Found [this one](https://www.reichelt.com/fr/fr/ci-unifilaire-to-92-ds-18b20-p58169.html?&trstct=pos_0&nbc=1) on reichlet.com, looks the same than [the one I bought from AZ-Delivery](https://www.amazon.fr/gp/product/B089QJ4TP5/ref=ppx_yo_dt_b_asin_title_o01_s00?ie=UTF8&psc=1), same price too in euros.
The one I bought looks genuine according to the lasered information even though it appears as counterfeit with discover_fake script. I don't know what to believe ^^
I will think about buying one piece from reichelt.com to check if the values are good.
And I think I will try to take the average of 10 readings to get 1 temperature. I want to make a temperature reading every 5 minutes outside, seems enough to keep track of weather.

Update after a few hours of the 2 boards running : 
the values are sensibly the same, sometimes exactly, sometimes +/- 0.1 or +/-0.01 of difference betwenn both, good enough to me.

**Comment by robert-hh** [comment]
I just confirmed that the DS18x20 and the actual ESP8266 code work. And yes, your code misses the wait time after convert().
